modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/store/Store.es.js

```diff
@@ -0,0 +1,256 @@
+			} finally {
+				isDispatching = false;
+			}
+
+			for (const subscriber of subscribers.values()) {
+				try {
+					subscriber();
+				} catch (error) {
+					console.error(error);
+				}
+			}
+		},
+
+		getState() {
+			if (isDispatching) {
+				throw new Error('Illegal getState() call during dispatch');
+			}
+
+			return state;
+		},
+
+		replaceReducer(nextReducer) {
+			currentReducer = nextReducer;
+
+			return store;
+		},
+
+		subscribe(subscriber) {
+			if (typeof subscriber !== 'function') {
+				throw new Error('`subscriber` must be a function');
+			}
+
+			if (isDispatching) {
+				throw new Error('Illegal subscribe() call during dispatch');
+			}
+
+			const nextSubscriberID = ++subscriberID;
+
+			subscribers.set(nextSubscriberID, subscriber);
+
+			return function unsubscribe() {
+				if (isDispatching) {
+					throw new Error(
+						'Illegal unsubscribe() call during dispatch'
+					);
+				}
+
+				subscribers.delete(nextSubscriberID);
+			};
+		},
+	};
+
+	return store;
+}
+
+/**
+ * Takes an object where the keys and values are namespaces and their
+ * corresponding reducer functions, and returns a single reducer function.
+ *
+ * See: https://redux.js.org/api/combinereducers
+ */
+export function combineReducers(reducers) {
+	const namespacesAndReducers = Object.entries(reducers);
+
+	return function combinedReducer(state, action) {
+		let hasChanged = false;
+
+		const nextState = {};
+
+		for ([namespace, reducer] of namespacesAndReducers) {
+			const previous = state[namespace];
+			const next = reducer(previous, action);
+
+			if (typeof next === 'undefined') {
+				throw new Error(`Reducer for ${namespace} returned undefined`);
+			}
+
+			hasChanged = hasChanged || previous !== next;
+
+			nextState[namespace] = next;
+		}
+
+		return hasChanged ? nextState : state;
+	};
+}
+
+export function defaultReducer(state, _action) {
+	return state;
+}
+
+/**
+ * Takes an array of reducers and returns a single reducer that runs them
+ * serially.
+ *
+ * This produces a "flat" state, unlike `combineReducers`, which produces a
+ * "nested/namespaced" state.
+ *
+ * See: https://github.com/redux-utilities/reduce-reducers
+ */
+export function reduceReducers(reducers) {
+	return function reducedReducer(state, action) {
+		return reducers.reduce((currentState, reducer, DELETE_ME) => {
+			const nextState = reducer(currentState, action);
+
+			if (typeof nextState === 'undefined') {
+				throw new Error(`Reducer for returned undefined`);
+			}
+
+			return nextState;
+		}, state);
+	};
+}
+
+const STORES = {};
+
+const SUB_QUEUE = {};
+
+export default {
+	Util: {
+		combineReducers,
+		reduceReducers,
+	},
+
+	get(name) {
+		return STORES[name];
+	},
+
+	subscribe(name, subscriber) {
+		if (STORES[name]) {
+			STORES[name].subscribe(subscriber);
+		} else {
+			if (SUB_QUEUE[name]) {
+				SUB_QUEUE[name].push(subscriber);
+			} else {
+				SUB_QUEUE[name] = [subscriber];
+			}
+		}
+	},
```

Comment:
I added this here because we can't quite count on the store to actually exist before subscribing to it. In the example above, the JS in the JSP actually gets run before the react component does, this means that if we don't have a mechanism like this, the store will always return undefined and unable to subscribe since it is run before the store is `.register`'d in the hook.

Reply:
It would be nice, but idk if it's possible or makes sense in the context of portal with many modules that can or can't exist on a given page. For example, say you have Portlet `A` & `B`. Let's day `A` is subscribed to a store that `B` creates. However, the `B` portlet is only conditionally on the page. I think it would be overeager of the store to make sure `B` is there and always registered first before `A`.

I think the mechanism above is helpful in the case that once the `B` portlet is on the page, `A` would automatically still be subscribed an not have to worry about it being registered or not. `A` would only have to worry about the data it gets from the store, and in the case where `B` is not registered yet, I would only need to know that no data exists yet.

Reply:
I am not sure that this addition makes for a coherent API, in the sense that on a Redux-compatible store, you have exactly these methods:

- `dispatch()`
- `getState()`
- `replaceReducer()`
- `subscribe()`

So, it seems a bit arbitrary to hoist exactly one of those methods up above the store layer and expose it here with a different signature (ie. adding `name` to it). It means we have near-duplication of a method at two layers, but the other two things you most obviously might want to do (eg. `dispatch()` and `getState()`) aren't possible at the upper layer.

If we wanted to hoist those up too, for consistency, we'd need to think about the semantics we want to define. For example, given that events can fire in any order (eg. store creation, subscription, dispatch), would we want "pending" dispatches to be held until creation? Until first subscription? Just dropped until there is a listener to actually do anything with them? Not a complete list there, but a sample of the kinds of scenarios we'd want to think about.